UDPServer/MainWindow.xaml.cs

```diff
@@ -0,0 +1,111 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Configuration;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using System.Windows;
+using System.Windows.Controls;
+using System.Windows.Data;
+using System.Windows.Documents;
+using System.Windows.Input;
+using System.Windows.Media;
+using System.Windows.Media.Imaging;
+using System.Windows.Navigation;
+using System.Windows.Shapes;
+
+namespace UDPServer
+{
+    /// <summary>
+    /// Interaction logic for MainWindow.xaml
+    /// </summary>
+    public partial class MainWindow : Window
+    {
+
+        MyUDPServer server = new MyUDPServer();
+        Thread listeningThread;
+
+        public MainWindow()
+        {
+            InitializeComponent();
+            ListBox1.ItemsSource = server.MessageReceived;
+            this.DataContext = server;
+            IPBox.Text = ConfigurationManager.AppSettings["IPAddress"];
+            PortBox.Text = ConfigurationManager.AppSettings["PortNumber"];
+        }
+
+        private void IPChanged(object sender, TextChangedEventArgs e)
+        {
+            server.ConnectionIP = IPBox.Text;
+        }
+
+        private void PortChanged(object sender, TextChangedEventArgs e)
+        {
+            server.ConnectionPort = Int32.Parse(PortBox.Text);
+        }
+
+        private void ConnectButtonClicked(object sender, RoutedEventArgs e)
+        {
+            if(!server.ConnectionStatus)
+            {
+                server.Connect();
+                listeningThread = new Thread(new ThreadStart(GetData));
```

Comment:
Replace Thread with `async/await`.

Use an async void ConnectButtonClicked to execute GetData().

Get Data() should also be an `async Task` method **without** the `while(true)`

Reply:
You may need a Task.Run with a cancellation token when you want to stop listening.

Reply:
Implemented